File: Week 03/id_306/FindMinimumInRotatedSortedArray.java
/**
 * Created by LynnSun on 2019/11/3.
 * 力扣题目地址：https://leetcode-cn.com/problems/find-minimum-in-rotated-sorted-array/
 */
public class FindMinimumInRotatedSortedArray {
    public int findMin(int[] nums) {
        int left = 0, right = nums.length - 1;
        while(left < right){
            //找到数组的中间元素 mid。
            int mid = (left + right) / 2;
            /*如果中间元素 > 数组第一个元素，变化点在右边。
            如果中间元素 < 数组第一个元素，变化点在左边*/
            if(nums[mid] > nums[right]) left = mid + 1;
            else right = mid;
        }
        return nums[left];
    }
}

File: Week 03/id_011/find-minimum-in-rotated-sorted-array.go
package algorithm00401

func findMin(nums []int) int {
	l, r := 0, len(nums)-1
	for l < r {
		mid := l + (r-l)/2
		if nums[l] >= nums[mid] && nums[mid] >= nums[r] {
			return nums[r]
		} else if nums[mid] > nums[r] {
			l = mid + 1
		} else if nums[l] > nums[mid] {
			r = mid
		} else {
			return nums[l]
		}
	}
	return nums[l]
}
